### PR TITLE
[WIP] Flax BLOOM implementation + demo

### PR DESCRIPTION
**Pull request by haileyschoelkopf**
# What does this PR do?

This PR will add a Flax implementation of BLOOM, and also I'd be happy to help contribute a tutorial / showcase of how to fine-tune BLOOM as well as discussed in #17703 :)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. --> linked above
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation). --> documentation in progress 
- [ ] Did you write any new necessary tests? --> will add once code is closer to completion

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@patrickvonplaten 

and @sanchit-gandhi @patil-suraj I believe were interested in collaborating. happy to discuss how best to do this.

### CODE REVIEW

**Comment by haileyschoelkopf** [comment]
A note on the initial status of this PR: (currently writing comment, will edit with more detail)

File: src/transformers/models/bloom/modeling_flax_bloom.py [hunk 1]
```diff
@@ -0,0 +1,769 @@
...
+        # TODO: init_cache is separate from use_cache flag
+        if init_cache:
+            outputs = (output,) + outputs
+        else:
+            outputs = (output,) + outputs[1:]
+
+        return outputs
+
+# TODO: does this still require position_ids?
+# TODO: gradient checkpointing
```
Comment (sanchit-gandhi) [inline src/transformers/models/bloom/modeling_flax_bloom.py:406]: For gradient checkpointing details, see: https://github.com/huggingface/transformers/issues/17399
Happy to lend a hand here! Have gotten it working myself in a Flax Seq2Seq Speech project
https://github.com/sanchit-gandhi/seq2seq-speech/blob/b28d0c25c8fad0f9ffa6707f91f7aba320d44a4b/models/modeling_flax_wav2vec2.py#L504

File: src/transformers/models/bloom/modeling_flax_bloom.py [hunk 2]
```diff
@@ -0,0 +1,769 @@
...
+
+        # TODO: init_cache is separate from use_cache flag
+        if init_cache:
+            outputs = (output,) + outputs
+        else:
+            outputs = (output,) + outputs[1:]
+
+        return outputs
+
+# TODO: does this still require position_ids?
```
Comment (sanchit-gandhi) [inline src/transformers/models/bloom/modeling_flax_bloom.py:405]: For Flax, we have a design philosophy in which we compute the `position_ids` and losses outside the model, and so need to pass the `position_ids` to the model!

**Review by sanchit-gandhi — COMMENTED** [review]
Looks like a great start @haileyschoelkopf, and a good set of TODO's for the next steps! Feel free to ping me (or @patrickvonplaten or @patil-suraj) if you encounter any diffculties or want to discuss ideas, very happy to help with the integration here!

**Comment by haileyschoelkopf** [comment]
Thanks for the helpful comments, @sanchit-gandhi ! I'll do another revision through the code fixing these and adding some more things as soon as I have time.

I think that the main thing that would need to be discussed soon is how to handle [AliBi](https://arxiv.org/abs/2108.12409) for position information, since it means that there is no specific max length for BLOOM inputs. I'm not too sure yet how to account for this given things like line 176, where the causal mask is made at the max length of the model and then sliced to get the mask for shorter sequences. (One idea I had was selecting a reasonable "starting max length", then if the model gets a longer input sequence the causal mask is extended either permanently or just for that forward pass).

File: src/transformers/models/bloom/modeling_flax_bloom.py [hunk 3]
```diff
@@ -0,0 +1,769 @@
...
+"""
+
+
+class FlaxBloomAttention(nn.Module):
+    config: BloomConfig
+    layer_number: int = None
+    dtype: jnp.dtype = jnp.float32
+
+    def setup(self):
+        config = self.config
```
[the remark this answers is not in the document]
Comment (haileyschoelkopf) [reply src/transformers/models/bloom/modeling_flax_bloom.py:126]: thanks, think the new commit will fix this and make it consistent!

**Comment by sanchit-gandhi** [comment]
Would there be any issues in implementing it in the first of the two ways proposed (set to `max_length`, slicing as required)?

The problem I envision with the latter of the two approaches is that once the function is jit'd, providing a new input length to the model would result in XLA having to recompile. Each time XLA sees a new input shape, it needs to trace out (compile) the function again. So if we provide a new input shape for each forward pass, XLA will recompile every time (very slow)! The performance benefits of jit'ing a function come when we re-use a function that has already been jit'd, meaning we should try and use fixed input shapes where possible.

**Comment by haileyschoelkopf** [comment]
Yeah, the recompilation is definitely something to try to avoid! 

But the issue is that [the bigscience/bloom config](https://huggingface.co/bigscience/bloom/blob/main/config.json) doesn't have any seq_length attribute ([but bigscience/bloom-1b3 does--4096](https://huggingface.co/bigscience/bloom-1b3/blob/main/config.json)) and we want BLOOM to be able to handle sequences as long as a user wants since AliBi allows generalization to longer sequences. We could maybe just choose a reasonable default `max_length`, and then if the user passes a sequence that's too long, permanently double the size of the causal mask--this would allow for fewer recompilations, hopefully.

But I think we should keep the possibility open to using the model on very long sequences without problems--I don't know if any other models in Transformers use AliBi embeddings yet so that's a unique benefit of this model.

File: src/transformers/models/bloom/modeling_flax_bloom.py [hunk 4]
```diff
@@ -0,0 +1,769 @@
...
+                f"`num_heads`: {self.num_heads})."
+            )
+
+        # Layer-wise attention scaling
+        self.norm_factor = jnp.sqrt(self.head_dim).astype(self.dtype) * self.layer_number
+        
+
+        self.attn_dropout = nn.Dropout(config.attention_dropout)
+
+        # Scaled Softmax TODO: change this to something implemented in jax (maybe implement in __call__ for attn module?)
```
[the remark this answers is not in the document]
Comment (haileyschoelkopf) [reply src/transformers/models/bloom/modeling_flax_bloom.py]: This class has an initial implementation now!

**Comment by haileyschoelkopf** [comment]
If just generating the causal mask at every forward pass is acceptable and wouldn't incur a speed penalty, then that should work fine!

And yes, I don't think that we need to pass position_ids into the model, and we can just compute the alibi embedding within the forward pass (the pytorch implementation does this.)

sorry for the delay on this--I'll work on it in the next 2 days.

**Comment by patrickvonplaten** [comment]
> If just generating the causal mask at every forward pass is acceptable and wouldn't incur a speed penalty, then that should work fine!
> 
> And yes, I don't think that we need to pass position_ids into the model, and we can just compute the alibi embedding within the forward pass (the pytorch implementation does this.)
> 
> sorry for the delay on this--I'll work on it in the next 2 days.

Great! Yeah, I just talked to @sanchit-gandhi offline - I think what we want to do here to only recompile when the model has to be recompiled anyways which translates into doing the folowing:

Allow `position_ids` to be passed but default them to `None` . If `None` they will be computed on the fly depending on the shape of `input_ids` and the values of `attention_mask` (the same would hold true for the causal_mask). Let me know if this doesn't make sense @haileyschoelkopf or if you have any other questions, more than happy to help :-)

**Comment by sanchit-gandhi** [comment]
Hey @haileyschoelkopf! This looks good with regards to the fused key-query-value matmuls in https://github.com/huggingface/transformers/pull/17761/commits/faddb8d446bfc8db0c1c77f29d9847a25f70b5a2! Just as a heads-up, for gradient checkpointing, you can follow the PR at https://github.com/huggingface/transformers/pull/17843. Feel free to reach out if there's anything you wish to discuss, very happy to help with any questions!

**Comment by sanchit-gandhi** [comment]
Hey @haileyschoelkopf! This looks good with regards to the fused key-query-value matmuls in https://github.com/huggingface/transformers/pull/17761/commits/faddb8d446bfc8db0c1c77f29d9847a25f70b5a2! Just as a heads-up, for gradient checkpointing, you can follow the PR at https://github.com/huggingface/transformers/pull/17843. Feel free to reach out if there's anything you wish to discuss, very happy to help with any questions!

**Comment by sanchit-gandhi** [comment]
Thank you @haileyschoelkopf for jumping on this so quickly and getting the structure for the model in place! This PR was completed in https://github.com/huggingface/transformers/pull/18022

Let me know if there's anything else you'd like to have a go at adding in JAX/Flax! Or if you'd like to have a go at porting another model to JAX/Flax I can make some suggestions!

**Comment by haileyschoelkopf** [comment]
Thanks so much for all the helpful comments @sanchit-gandhi on this PR and apologies I wasn't able to iterate quicker on it!

If I have more time to add another JAX model I'll ping you for sure :)

**Comment by haileyschoelkopf** [comment]
Of course, will ping you if so :)
